ws-server/src/server.rs

```diff
@@ -21,477 +21,201 @@
-	SendOutNotif {
-		/// The value that was passed in [`FrontToBack::RegisterSubscription::unique_id`] earlier.
-		unique_id: usize,
-		/// Notification to send to the subscribed clients.
-		notification: JsonValue,
-	},
+		Ok(())
+	}
 }
 
 impl Server {
-	/// Initializes a new server.
-	pub async fn new(url: impl AsRef<str>) -> Result<Self, Box<dyn error::Error + Send + Sync>> {
-		let sockaddr: SocketAddr = url.as_ref().parse()?;
-		let transport_server = WsTransportServer::builder(sockaddr).build().await?;
-		let local_addr = *transport_server.local_addr();
-
-		// We use an unbounded channel because the only exchanged messages concern registering
-		// methods. The volume of messages is therefore very low and it doesn't make sense to have
-		// a backpressure mechanism.
-		// TODO: that's not true anymore ^
-		let (to_back, from_front) = mpsc::unbounded();
-
-		async_std::task::spawn(async move {
-			background_task(transport_server.into(), from_front).await;
-		});
-
-		Ok(Server {
-			to_back,
-			registered_methods: Arc::new(Mutex::new(Default::default())),
-			next_subscription_unique_id: Arc::new(atomic::AtomicUsize::new(0)),
-			local_addr,
-		})
+	pub fn register_method<F, R>(&mut self, method_name: &'static str, callback: F)
+	where
+		R: Serialize,
+		F: Fn(&str) -> Result<R, RpcError> + Send + Sync + 'static, // TODO: figure out correct lifetime here
+	{
+		self.methods.insert(
+			method_name,
+			Box::new(move |id, params, tx, _| {
+				let result = callback(params)?;
+
+				let json = serde_json::to_string(&JsonRpcResponse { jsonrpc: TwoPointZero, id, result })?;
+
+				tx.send(json).map_err(Into::into)
+			}),
+		);
 	}
 
-	/// Local socket address of the underlying transport server.
-	pub fn local_addr(&self) -> &SocketAddr {
-		&self.local_addr
-	}
+	pub fn register_subscription<T>(
+		&mut self,
+		subscribe_method_name: &'static str,
+		unsubscribe_method_name: &'static str,
+	) -> SubsciptionSink {
+		let subscribers = Arc::new(Mutex::new(FxHashMap::default()));
 
-	/// Registers a notification method name towards the server.
-	///
-	/// Clients will then be able to call this method.
-	/// The returned object allows you to process incoming notifications.
-	///
-	/// If `allow_losses` is true, then the server is allowed to drop notifications if the
-	/// notifications handler (i.e. the code that uses [`RegisteredNotifications`]) is too slow
-	/// to process notifications.
-	///
-	/// Returns an error if the method name was already registered.
-	pub fn register_notification(
-		&self,
-		method_name: String,
-		allow_losses: bool,
-	) -> Result<RegisteredNotification, Error> {
-		if !self.registered_methods.lock().insert(method_name.clone()) {
-			return Err(Error::MethodAlreadyRegistered(method_name));
-		}
+		{
+			let subscribers = subscribers.clone();
+			self.methods.insert(
+				subscribe_method_name,
+				Box::new(move |id, _, tx, conn| {
+					let sub_id = {
+						const JS_NUM_MASK: u64 = !0 >> 11;
```

Comment:
This is used because [JavaScrip` max number is 2^53-1](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/MAX_SAFE_INTEGER)?